### PR TITLE
IoT agent refactor

### PR DESCRIPTION
**Pull request by giovannicuriel**
* **Please check if the PR fulfills these requirements**
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Refactoring

* **What is the current behavior?** (You can also link to an open issue here)
IoT agent mosca is virtually a single file application.

* **What is the new behavior (if this is a feature change)?**
All internal modules were properly separated. There are a few things to improve, though, but this might be a good start.

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
No.

* **Is there any issue related to this PR in other repository?** (such as dojot/dojot)
No.

* **Other information**:

### CODE REVIEW

**Comment by giovannicuriel** [comment]
A few things:
- As you can see, I tried to separate things into reusable and replaceable components - for instance: if needed we could change MqttBacked to use the infamous aedes MQTT implementation with no changes to other components (or, at least, changes are kept to a minimum)
- I tried to move the metrics from development branch to this refactor, but I could not find where the code gathers all connections by minute, 5 minutes and so on. I might have missed something - could you check that?
- Device authorization is still messy. 
- Tests are missing. I had a branch with pretty good test coverage, but everything changed since then, and I got lazy enough to not rewrite them right now ;)

File: package-lock.json [hunk 1]
```diff
@@ -4259,39 +4264,39 @@
...
     },
     "uuid": {
       "version": "3.3.2",
-      "resolved": "https://registry.npmjs.org/uuid/-/uuid-3.3.2.tgz",
+      "resolved": "http://localhost:4873/uuid/-/uuid-3.3.2.tgz",
       "integrity": "sha512-yXJmeNaw3DnnKAOKJE51sL/ZaYfWJRl1pK9dr19YFCu0ObS231AB1/LbqTKRAQ5kw8A90rA6fr4riOUpTZvQZA=="
     },
     "vary": {
       "version": "1.1.2",
-      "resolved": "https://registry.npmjs.org/vary/-/vary-1.1.2.tgz",
```
Comment (mprevide) [inline package-lock.json]: The correct would not be registry.npmjs.org for all?

File: src/mosca.js [hunk 2]
```diff
@@ -0,0 +1,485 @@
...
+        db: 12,
+        port: config.backend_port,
+        return_buffers: true, // to handle binary payloads
+        host: config.backend_host
+      },
+      persistence: {
+        factory: mosca.persistence.Redis,
+        host: config.backend_host
+      },
+      interfaces: moscaInterfaces,
```
Comment (nunesjoab) [inline src/mosca.js]: I think the parameter responsible for provide Mosca metrics (stats: true) is missing, corresponding to line 107 of the index.js file of branch development

File: src/mosca.js [hunk 3]
```diff
@@ -0,0 +1,485 @@
+"use strict";
+const mosca = require("mosca");
+const config = require("./config");
+const logger = require("@dojot/dojot-module-logger").logger;
+const util = require("util");
+
+const TAG = { filename: "mqtt-backend"};
+
```
Comment (nunesjoab) [inline src/mosca.js]: At this point we should define a variable that receives the logger level passed through environment variable by docker-compose and set logger level using function logger.setLevel()

**Comment by giovannicuriel** [comment]
@raulnegreiros As I remember, the default connection method would be using TLS. If the user really wants it, a plain connection could be enabled (there would not be a way to disable secured connections, though).

AFAIK, the code that communicates with EJBCA and generate the keys and certificates was shamelessly copied from auth. We could improve on that :D

File: package-lock.json (hunk 1 above)
Comment (giovannicuriel) [reply package-lock.json]: I don't know what happened, actually. Yes, it should be npmjs.org.

File: src/mosca.js (hunk 2 above)
Comment (giovannicuriel) [reply src/mosca.js]: Done!

File: src/mosca.js (hunk 3 above)
Comment (giovannicuriel) [reply src/mosca.js]: Done!

File: entrypoint.sh [hunk 4]
```diff
@@ -1,13 +1,13 @@
 #!/bin/sh
 
 echo "Generating keys and certificates for TLS ..."
-/opt/iot-agent/initialConf.py && echo "TLS configured." || { echo "Error on TLS setup"; exit 1; }
+./initialConf.py && echo "TLS configured." || { echo "Error on TLS setup"; exit 1; }
```
Comment (jonaphael) [inline entrypoint.sh]: when running this python script on alpine based container, an undefined reference throws.

solution: python ./initialConf.py

**Comment by nunesjoab** [comment]
> A few things:
> 
> * As you can see, I tried to separate things into reusable and replaceable components - for instance: if needed we could change MqttBacked to use the infamous aedes MQTT implementation with no changes to other components (or, at least, changes are kept to a minimum)
> * I tried to move the metrics from development branch to this refactor, but I could not find where the code gathers all connections by minute, 5 minutes and so on. I might have missed something - could you check that?
> * Device authorization is still messy.
> * Tests are missing. I had a branch with pretty good test coverage, but everything changed since then, and I got lazy enough to not rewrite them right now ;)
@giovannicuriel I'm developing Unit Tests and would be a great help if you could commit this branch with tests. Is it possible?
> 
> It may be worth noting that I was able to execute this IoT agent using NodeJS v10. I tried to use v12 but there were some weird compilation issues while building C bindings from nan and node-rdkafka packages.

**Comment by giovannicuriel** [comment]
@joab-cpqd Sure, I'll add them. I think I'll have time this week to do so (including figuring out a way to generate certificates without python code)

File: entrypoint.sh (hunk 4 above)
Comment (giovannicuriel) [reply entrypoint.sh]: I'll try to add code in IoT agent to generate these certificates from inside the code (as pointed out @raulnegreiros - furthermore, we have control of whether this procedure is executed or not from within the code, which might be a good thing IMO)

**Comment by nunesjoab** [comment]
iot

> @nunesjoab Sure, I'll add them. I think I'll have time this week to do so (including figuring out a way to generate certificates without python code).
Thanks! :smile: 
> 
> BTW, is your user @joab-cpqd or @nunesjoab?
I created @joab-cpqd when I started working on Dojot, but my main user was always @nunesjoab, so I switched to the ex and I'm going to use it now

**Comment by jonaphael** [comment]
Hi, @giovannicuriel thank you so much for this PR, we appreciate your help with this refactor and it helps a lot, so the team will merge this PR, then "fix" the comments, no problem with that? Another point is: have you made some progress on this code and will push them soon? Or can we wait until the next changes before merging?

**Comment by giovannicuriel** [comment]
Hi, @jonaphael! I don't know which would be the best approach. I was planning to work on this PR this week, but I'm not sure if I'll have time.
I don't have any complains related to merge it now and add unit tests later :wink:

**Comment by giovannicuriel** [comment]
@jonaphael  I'd fix the issue you pointed out before merging, though (the one in which python was not in alpine image). 

(When coding and testing at home, I don't use docker because of... reasons)

**Pull request merged**